### ISSUE TITLE
systemd unit unable to start properly

### ISSUE THREAD

**Issue by TheAwoo** [mdgaziur/findex#27]
Hi,
The current version installed using [aur](https://aur.archlinux.org/packages/findex-bin) fails to boot under systemd:

```
systemd[1187]: Started Findex.
findex[14968]: [ERROR] Only X11 is supported
systemd[1187]: findex.service: Main process exited, code=exited, status=1/FAILURE
systemd[1187]: findex.service: Failed with result 'exit-code'.
systemd[1187]: Stopped Findex.
```
The program works fine when ran manually or added to session startup.

**Comment by mdgaziur** [comment]
Did you start it with `systemctl --user findex.service`?

**Comment by TheAwoo** [comment]
Yes, I'm using the usual commands to manage the unit:

```
awoo@~ $ systemctl --user start findex
awoo@~ $ systemctl --user stop findex
```

I've also decided to add some additional logging and confirmed, using a locally built version, that the variable is unavailable only when running from systemd:

```
systemd[1352]: Started Findex.
findex[27702]: thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NotPresent', src/main.rs:15:57
findex[27702]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
systemd[1352]: findex.service: Main process exited, code=exited, status=101/n/a
systemd[1352]: findex.service: Failed with result 'exit-code'.
systemd[1352]: Stopped Findex.
```

Modified main.rs:
```rust
if std::env::var("XDG_SESSION_TYPE") != Ok(String::from("x11")) {
    let session = std::env::var("XDG_SESSION_TYPE").unwrap();
    eprintln!("[ERROR] Only X11 is supported, got session: {}", session);
    std::process::exit(1);
}
```

**Comment by mdgaziur** [comment]
Weird, it does work in my machine because the variables are set. Try to remove the check and see if it works.

**Comment by TheAwoo** [comment]
The application now launches only when I manually start it after logging in, otherwise it's dead on session startup.
Logs are spotty at best, here's what I've been able to capture:

```
systemd[765]: Started Findex.
findex[1652]: [WARN] Unknown XDG session type!
findex[1652]: [INFO] Starting Findex...
findex[1652]: thread 'main' panicked at 'Failed to init GTK: BoolError { message: "Failed to initialize GTK", filename: "/home/awoo/.cargo/registry/src/github.com-1ecc6299db9ec823/gtk-0.15.5/src/rt.rs", function: "gtk::rt", line: 134 }', src/main.rs:19:17
findex[1652]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
systemd[765]: findex.service: Main process exited, code=exited, status=101/n/a
systemd[765]: findex.service: Failed with result 'exit-code'.
systemd[765]: findex.service: Scheduled restart job, restart counter is at 2.
systemd[765]: Stopped Findex.
```

**Comment by mdgaziur** [comment]
Does systemd show any error or warning about the service file? It should
start after graphical session starts

**Comment by TheAwoo** [comment]
No, all I see in boot logs is the usual info and logs implying that the app has started okay, binding to hotkeys and all that.
Why it's dead when checking it after login remains a mystery (no logs whatsoever).

**Comment by mdgaziur** [comment]
At this moment, I also don't have any clue either. Probably there's
something wrong with findex.service or something on your side

**Comment by TheAwoo** [comment]
Reckon it's a weird side effect of my setup as both my computers running arch are affected.
Thankfully adding the app to xfce session autostart works just fine so I'll stick with that. :)

**Comment by mdgaziur** [comment]
Can I close the issue?

**Comment by TheAwoo** [comment]
Sure thing. Thank you for your assistance!

**Comment by trinitronx** [comment]
> The application now launches only when I manually start it after logging in, otherwise it's dead on session startup. Logs are spotty at best, here's what I've been able to capture:
> 
> ```
> systemd[765]: Started Findex.
> findex[1652]: [WARN] Unknown XDG session type!
> findex[1652]: [INFO] Starting Findex...
> findex[1652]: thread 'main' panicked at 'Failed to init GTK: BoolError { message: "Failed to initialize GTK", filename: "/home/awoo/.cargo/registry/src/github.com-1ecc6299db9ec823/gtk-0.15.5/src/rt.rs", function: "gtk::rt", line: 134 }', src/main.rs:19:17
> findex[1652]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> systemd[765]: findex.service: Main process exited, code=exited, status=101/n/a
> systemd[765]: findex.service: Failed with result 'exit-code'.
> systemd[765]: findex.service: Scheduled restart job, restart counter is at 2.
> systemd[765]: Stopped Findex.
> ```


@mdgaziur, @TheAwoo: It turns out that I'm able to reproduce this error.

Recently I noticed that there were many gigabytes of log files in `~/.findex-logs/`.  They all had this same message repeated:

> ```
> thread 'main' panicked at crates/findex/src/main.rs:25:17:
> Failed to init GTK: BoolError { message: "Failed to initialize GTK", filename: "/home/exampleuser/.cargo/registry/src/index.crates.io-6f17d22bba15001f/gtk-0.15.5/src/rt.rs", function: "gtk::rt", line: 134 }
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> [WARN] Findex exited unexpectedlywith exit code: 101
> [INFO] Respawning Findex
> thread 'main' panicked at crates/findex/src/main.rs:25:17:
> Failed to init GTK: BoolError { message: "Failed to initialize GTK", filename: "/home/exampleuser/.cargo/registry/src/index.crates.io-6f17d22bba15001f/gtk-0.15.5/src/rt.rs", function: "gtk::rt", line: 134 }
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> [WARN] Findex exited unexpectedlywith exit code: 101
> [INFO] Respawning Findex
> 
> [...SNIP...] (Message repeats for over 17 Gigabytes worth of text!!)
> ```


Seems that the error is from the `gtk` crate.  I can't find any related issues upstream and [the repo has been archived][1] (says now unmaintained?).

Anyway, looking at `gtk` version `0.15.5` I can see that it's from [this line here][2].

It's the error case inside `gtk::init()`, handling when the following `if` statement returns `false`:

```rust
if from_glib(ffi::gtk_init_check(&mut 1, &mut argv.to_glib_none().0)) {
    // See https://github.com/gtk-rs/gtk-rs-core/issues/186 for reasoning behind
    // acquiring and leaking the main context here.
    let result: bool = from_glib(glib::ffi::g_main_context_acquire(
        glib::ffi::g_main_context_default(),
    ));
    if !result {
        return Err(glib::bool_error!("Failed to acquire default main context"));
    }
    set_initialized();
    Ok(())
} else {
    Err(glib::bool_error!("Failed to initialize GTK"))
}
```

Not sure how to dig further, or if it's worth trying to upgrade to latest `gtk` crate.  It's doing some FFI call out to the GTK C code in an `unsafe` block.  This would be `extra/gtk3` version `1:3.24.48-1` on Arch Linux-based distros.

They recommend moving to `gtk4-rs` now, but that probably entails more changes.

[1]: https://github.com/gtk-rs/gtk3-rs/
[2]: https://github.com/gtk-rs/gtk3-rs/blob/51df5c22fbe02ec57c5052a3e5a91cc9671a0e83/gtk/src/rt.rs#L134

**Comment by mdgaziur** [comment]
That's pretty interesting. Can you try creating a findex user unit instead? Let me know whether that fixes the issue. 

Also, I think it'll be a good idea to add a file size limit to the log files to prevent horribly large log files. I'll add that soon after the admission exams end.

I'm a bit busy with academia right now so I can't test it on my system.

**Comment by trinitronx** [comment]
@mdgaziur:

> Can you try creating a findex user unit instead?

This was already with using `findex-daemon.service` as SystemD user unit:

```desktop
# ~/.config/systemd/user/findex-daemon.service
[Unit]
Description=Findex Daemon
PartOf=graphical-session.target
After=graphical-session.target

[Service]
Type=forking
ExitType=main
ExecStart=/usr/bin/findex-daemon
Restart=on-failure

[Install]
WantedBy=graphical-session.target
```

The large log size was from the automatic restarts (e.g. a "crashloop"), since it kept restarting and crashing with that error.

To avoid that log file size build-up we could try adding a restart limit to the `[Service]` section:

```desktop
[Service]
### ... The above usual parts
### [... SNIP ...]
# Exponential CrashLoop Backoff
RestartSteps=5
RestartMaxDelaySec=60s
# Give up after 6 tries
StartLimitInterval=infinity
StartLimitBurst=6
```

That should prevent the infinite restart crashloop, and limit log size to 6x the per-restart log output size.

The other thing we could do to limit log size in general is to use log rotation & file size limiting.  If we're already using SystemD, then we can benefit from the [bult-in JournalD limits][1] 

<details><summary>Expand for JournalD settings details (configured in <code>/etc/systemd/journald.conf`</code>:</summary>
<p>

- `SystemMaxUse`: Controls how much disk space the journal may use up at most. (default `10%` of current filesystem, capped to `4Gib`)
- `SystemKeepFree`: Controls how much disk space `systemd-journald` shall leave free for other uses.  This attempts to keep this much space on the filesystem free, but in some cases this value can be exceeded.
  - Although, this is evaluated at startup and the way it works can lead to more space being used than expected.
  - JournalD docs state:
    - > If the file system is nearly full and either `SystemKeepFree=` or `RuntimeKeepFree=` are violated when `systemd-journald` is started, the limit will be raised to the percentage that is actually free.
    - > This means that if there was enough free space before and journal files were created, and subsequently something else causes the file system to fill up, journald will stop using more space, but it will not be removing existing files to reduce the footprint again, either.
    - > Also note that only archived files are deleted to reduce the space occupied by journal files.
    - > This means that, in effect, there might still be more space used than `SystemMaxUse=` or `RuntimeMaxUse=` limit after a vacuuming operation is complete.
- `SystemMaxFileSize`: Controls how large individual journal files may grow at most. (default `1/8 * SystemMaxUse`, capped to `128Mib` so that usually seven rotated journal files are kept as history)
- `SystemMaxFiles`: Controls how many individual journal files to keep at most. (default `100` files)
  - In some cases, there may still be more files left around than this limit.
  - JournalD docs state:
    - > Note that only archived files are deleted to reduce the number of files until this limit is reached; active files will stay around.
    - > This means that, in effect, there might still be more journal files around in total than this limit after a vacuuming operation is complete.

</p>
</details>

To use JournalD, all that needs to be done is log directly to stdout.  Note that this depends on the user enabling the SystemD user unit for `findex-daemon.service` as the startup method.

In other cases, for daemons that log to files instead of directly to JournalD, (e.g. `~/findex-logs/*.log`) usually the best-practice is to setup a `logrotate.d` configuration to rotate & compress old log files and to delete the oldest ones when they get "rotated" out by new ones.

For example, a `logrotate.d` config might look like:

```Nginx
/var/log/findex/*.log {
  daily
  copytruncate
  missingok
  compress
  rotate 7
  maxsize 512k
  dateformat -%Y%m%d-%s
  postrotate
    /etc/init.d/findex-daemon restart
  endscript
}
```

Note that this example would need to be adjusted if `findex-daemon` is logging to some other place than `/var/log/findex/`, and the `postrotate` would only be for users who use SysV Init.  The `postrotate` for restarting `findex-daemon` might be different when launching with Sway, etc...  (Maybe something like: `pkill findex-daemon; swaymsg exec /usr/bin/findex-daemon`)


[1]: https://www.freedesktop.org/software/systemd/man/latest/journald.conf.html#SystemMaxUse=

**Comment by mdgaziur** [comment]
The systemd service is proving to be quite troublesome. Findex daemon mostly achieves what the systemd service does. So, is it really necessary though? Startup of the daemon can be done via the DE or wm the user is using if I'm not wrong.

**Comment by trinitronx** [comment]
> The systemd service is proving to be quite troublesome. Findex daemon mostly achieves what the systemd service does. So, is it really necessary though? Startup of the daemon can be done via the DE or wm the user is using if I'm not wrong.

While it is possible to launch a daemon with either method [^1] (e.g. SystemD or via DE/WM), it's generally more beneficial in cases like this where a daemon might be crashing to attempt to restart it.  There are benefits to automatic restarts that ensure higher reliability of services, even when some bugs or crashes are happening.  The caveat to this is that logs might be generated in this case filling up a file as a side-effect.  Regardless of which method we use to launch `findex-daemon`, this would be a problem eventually surfacing to the user.

It makes sense to limit the number of restarts if we are not going to change logging mode to point to JournalD when using the SystemD service launch method, especially given that this is a desktop-centric daemon rather than a production server daemon (e.g. webserver, API server, etc...).  If it were a server daemon, I'd more strongly suggest always logging to JournalD and prioritizing fixing the root cause of the crash, but since it's a desktop daemon, we can be a bit more flexible.

I'll go ahead and make a PR for the restart limit change for the SystemD service, noting that using SystemD is optional (e.g. the user must choose to run: `sudo systemctl enable --user findex-daemon.service` to use it).

**EDIT:** I've pushed up the recommended fix in mdgaziur/findex#72

[^1]: Just to be clear, typical `findex-daemon` crashes would impact users regardless of whether SystemD was used to launch the daemon, or not. In the case of SystemD, restarts would be attempted when the daemon crashes, while in the case of being launched via Sway, it would crash once and remain stopped.
